src/main/java/org/prebid/server/bidder/pangle/PangleBidder.java

```diff
@@ -0,0 +1,215 @@
+                final Integer adType = resolveAdType(imp, extBidder);
+                final Imp modifiedImp = modifyImp(imp, adType, extBidder);
+
+                requests.add(createRequest(request, modifiedImp, extBidder.getBidder().getToken()));
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        return Result.of(requests, errors);
+    }
+
+    private WrappedImpExtBidder parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), WrappedImpExtBidder.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(String.format("failed unmarshalling imp ext (err)%s", e.getMessage()));
+        }
+    }
+
+    private static Integer resolveAdType(Imp imp, WrappedImpExtBidder extBidder) {
+
+        if (imp.getVideo() != null) {
+            final ExtImpPrebid extPrebid = extBidder != null ? extBidder.getPrebid() : null;
+            final Integer isRewardedInventory = extPrebid != null ? extPrebid.getIsRewardedInventory() : null;
+            if (isEqualIntegers(isRewardedInventory, 1)) {
+                return 7;
+            }
+
+            if (isEqualIntegers(imp.getInstl(), 1)) {
+                return 8;
+            }
+        }
+
+        if (imp.getBanner() != null) {
+            if (isEqualIntegers(imp.getInstl(), 1)) {
+                return 2;
+            } else {
+                return 1;
+            }
+        }
+        if (imp.getXNative() != null && StringUtils.isNotBlank(imp.getXNative().getRequest())) {
+            return 5;
+        }
+
+        return -1;
+    }
+
+    private static boolean isEqualIntegers(Integer firstNum, Integer secondNum) {
+        return firstNum != null && firstNum.equals(secondNum);
+    }
+
+    private Imp modifyImp(Imp imp, Integer adType, WrappedImpExtBidder extBidder) {
+        if (adType == -1) {
+            throw new PreBidException("not a supported adtype");
+        }
+        final WrappedImpExtBidder updatedImpExt = extBidder.toBuilder().adType(adType).build();
+        return imp.toBuilder()
+                .ext(mapper.mapper().convertValue(updatedImpExt, ObjectNode.class))
+                .build();
+    }
+
+    private HttpRequest<BidRequest> createRequest(BidRequest request, Imp imp, String token) {
+        final BidRequest outgoingRequest = request.toBuilder().imp(Collections.singletonList(imp)).build();
+
+        return HttpRequest.<BidRequest>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(makeHeaders(token))
+                .payload(outgoingRequest)
+                .body(mapper.encode(outgoingRequest))
+                .build();
+    }
+
+    private static MultiMap makeHeaders(String token) {
+        return HttpUtil.headers()
+                .add("TOKEN", token);
```

Comment:
Can token be null ?

Reply:
From pangle.json
"required": [
    "token"
  ]
  Seems it can't be null